design/ceph/ceph-orchestrator-osd-add-remove.md

```diff
@@ -0,0 +1,266 @@
+specification. The parent Orchestrator Module in Ceph will not store user-specified Drive Groups;
+however, it does expect the underlying Orchestrator or Orchestrator Mgr Module to store the Drive
+Group spec.
+
+Even with Drive Groups, Ceph expects the on-disk content of disks to be the declarative state of
+whether an OSD belongs to a Ceph cluster or not and what its configuration is. Rook is currently
+able to read this on-disk content via `ceph-volume` to determine if it should run a Kubernetes
+Deployment for a disk-based OSD.
+
+Rook currently does not implement OSD removal via any declarations in the `CephCluster` resource in
+order to prioritize the safety of user data. If a user wishes to remove an OSD, they must perform
+manual, imperative steps to do so.
+
+
+## Proposal
+
+### Feature requests to Ceph project
+This proposal will require changes to Rook and to Ceph's Rook Orchestrator Mgr Module. These changes
+are assumed, but requested changes to other Ceph components will be noted here.
+
+1. `ceph-volume` - Accept Drive Groups as input via command line flag
+   1. Should be automatically converted and issued `ceph-volume lvm batch` commands under the hood
+   2. Support JSON and YAML strings as inputs to command line flag
+   3. If the node matching glob for the Drive Group does not apply for the current node,
+     `ceph-volume` should report success without taking any actions
+2. Drive Groups - support a blacklist of devices
+   1. Should be able to blacklist individual devices by the device path
+   2. The blacklist may be able to support other ways of specifying devices to blacklist, but above
+     is the minimum requirement
+
+### Adding OSDs
+This design proposes to add a section to the `CephCluster` resource's `storage` configuration
+section to allow disks to be added via a Drive Groups specification. The spec is proposed below with
+behavior described to follow.
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephCluster
+spec:
+  storage: # preexisting
+    driveGroups: # proposed new
+      - spec: < string, JSON or YAML blob to pass to ceph-volume for OSD creation >
+```
+
+Each individual `driveGroups` item's `spec` will be a simple string containing a JSON blob or YAML
+doc string for the Drive Group. `ceph-volume` feature request #1 will be used here to allow Rook to
+pass along the Drive Group JSON/YAML string without modification or inspection to `ceph-volume` when
+running the OSD provisioning job. OSD creation via Drive Groups is given priority and executed
+before any other CRD-based OSD creation methods are used (the other CRD-based creations methods are
+still used after the Drive Group method).
+
+Drive Groups should be defined only at the cluster level, not at node level. Drive Groups support a
+glob matching pattern for nodes, so every drive group should be issued on every node. Rook will not
+suspect an error for non-matching nodes due to `ceph-volume` feature 1-3.
```

Comment:
As we do not know what OSD's produce each drive group. Are we going to launch the OSD creation via Drive Groups in each reconcile loop?

Reply:
That is what this is stating, yes. Each Drive Group is issued on every host on every orchestration.

Reply:
Is not very nice ...